src/bun.js/bindings/NodeModuleModule.zig

```diff
@@ -81,3 +81,155 @@ pub fn _stat(path: []const u8) i32 {
+    if (kind >= 0) {
+        const loader: CustomLoader = switch (kind) {
+            1 => .{ .loader = .js },
+            2 => .{ .loader = .json },
+            3 => .{ .loader = .napi },
+            4 => .{ .loader = .ts },
+            else => .{ .custom = undefined }, // to be filled in later
+        };
+        const gop = try list.getOrPut(bun.default_allocator, str);
+        if (!gop.found_existing) {
+            const dupe = try bun.default_allocator.dupe(u8, str);
+            errdefer bun.default_allocator.free(dupe);
+            gop.key_ptr.* = dupe;
+            if (is_built_in) {
+                vm.has_mutated_built_in_extensions += 1;
+            }
+            gop.value_ptr.* = .pack(switch (loader) {
+                .loader => loader,
+                .custom => .{
+                    .custom = JSCommonJSExtensions__appendFunction(global, value),
+                },
+            });
+        } else {
+            const existing = gop.value_ptr.*.unpack();
+            if (existing == .custom and loader != .custom) {
+                swapRemoveExtension(vm, existing.custom);
+            }
+            gop.value_ptr.* = .pack(switch (loader) {
+                .loader => loader,
+                .custom => .{
+                    .custom = if (existing == .custom) new: {
+                        JSCommonJSExtensions__setFunction(global, existing.custom, value);
+                        break :new existing.custom;
+                    } else JSCommonJSExtensions__appendFunction(global, value),
+                },
+            });
+        }
+    } else if (list.fetchSwapRemove(str)) |prev| {
+        bun.default_allocator.free(prev.key);
+        if (is_built_in) {
+            vm.has_mutated_built_in_extensions -= 1;
+        }
+        switch (prev.value.unpack()) {
+            .loader => {},
+            .custom => |index| swapRemoveExtension(vm, index),
+        }
+    }
+}
+
+fn swapRemoveExtension(vm: *JSC.VirtualMachine, index: u32) void {
+    const last_index = JSCommonJSExtensions__swapRemove(vm.global, index);
+    if (last_index == index) return;
+    // Find and rewrite the last index to the new index.
+    // Since packed structs are sugar over the backing int, this code can use
+    // the simd path in the standard library search.
+    const find: u32 = @intFromEnum(CustomLoader.Packed.pack(.{ .custom = last_index }));
+    const values = vm.commonjs_custom_extensions.values();
+    const values_reinterpret = bun.reinterpretSlice(u32, values);
+    const i = std.mem.indexOfScalar(u32, values_reinterpret, find) orelse
+        return bun.debugAssert(false);
+    values[i] = .pack(.{ .custom = last_index });
+}
+
+pub fn findLongestRegisteredExtension(vm: *JSC.VirtualMachine, filename: []const u8) ?CustomLoader {
+    const basename = std.fs.path.basename(filename);
+    var next: usize = 0;
+    while (true) {
+        const i = bun.strings.indexOfCharPos(basename, '.', next) orelse
```

Comment:
This looks dangerous; a bug introduced later could cause an infinite loop. Consider enforcing an upper bound on # of iterations.

Reply:
what